### PR TITLE
feat: Collapse All Apex Tests functionality @W-16273375@

### PR DESCRIPTION
**Pull request by jamessimone**
### What does this PR do?

* Adds collapse all apex test functionality to testing side panel
* Updates default behavior for test methods to be collapsed - for projects with even a handful of test classes, the lack of a search bar becomes a big hindrance when using the test panel to scroll through
* Fixes a few linting errors by adding the appropriate types

Thanks to @gbockus-sf for pairing with me to tighten this PR up!

### What issues does this PR fix or reference?

@W-16273375@

### Functionality Before
image

### Functionality After
image

### CODE REVIEW

File: packages/salesforcedx-vscode-apex/src/index.ts [hunk 1]
```diff
@@ -343,17 +358,22 @@ const createLanguageClient = async (
       );
     }
   } catch (e) {
-    languageClientUtils.setStatus(ClientStatus.Error, e);
-    let eMsg =
-      typeof e === 'string' ? e : e.message ?? nls.localize('unknown_error');
+    const err = e as unknown as Error;
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex/src/index.ts:361]: Use of as unknown as ... is discouraged. I would rather see type guard

```
isDuration(
  predicateOrWait: PredicateWithTimeout | Duration
): predicateOrWait is Duration {
  return (predicateOrWait as Duration).milliseconds !== undefined;
}
```

File: packages/salesforcedx-vscode-apex/src/views/testOutlineProvider.ts [hunk 2]
```diff
@@ -316,12 +322,10 @@ export class ApexTestGroupNode extends TestNode {
     this.failing = 0;
     this.skipping = 0;
     this.children.forEach(child => {
-      if ((child as ApexTestNode).outcome === PASS_RESULT) {
-        this.passing++;
-      } else if ((child as ApexTestNode).outcome === FAIL_RESULT) {
-        this.failing++;
-      } else if ((child as ApexTestNode).outcome === SKIP_RESULT) {
-        this.skipping++;
+      if (child instanceof ApexTestNode) {
```
Comment (peternhale) [inline packages/salesforcedx-vscode-apex/src/views/testOutlineProvider.ts:325]: nice - thanks

**Review by peternhale — CHANGES_REQUESTED** [review]

File: packages/salesforcedx-vscode-apex/src/index.ts (hunk 1 above)
Comment (jamessimone) [reply packages/salesforcedx-vscode-apex/src/index.ts]: Updated! Thanks for the feedback

**Comment by jamessimone** [comment]
Apologies, [#7e69e45](https://github.com/forcedotcom/salesforcedx-vscode/pull/5684/commits/7e69e45f8df16cd83eaa4f727835f7d16491c9ab) snuck in while I was signing my gpg key

**Review by peternhale — CHANGES_REQUESTED** [review]
@jamessimone can you make initial state expanded?

**Comment by jamessimone** [comment]
@peternhale I've updated that. I will probably try to later contribute a piece of configuration to allow consumers to set the default state for tests, though, as expanded by default makes for a lot of scrolling in a project with more than a few tests, and particularly when there are more than a few tests in the first test class shown.

Here are a few examples using 2560x1440 resolution:

- [Apex Rollup](https://github.com/jamessimone/apex-rollup): shows 2.5 test classes with all tests expanded, shows all 30 test classes collapsed 
- [Nebula Logger](https://nebulalogger.com/): shows nearly 7 test classes expanded, shows 57 out of 65 tests collapsed
- Internal Salesforce.com repo 1: shows 4 test classes expanded, shows all 24 test classes collapsed
- Internal Salesforce.com repo 2: shows 2 test classes expanded, shows all 18 test classes collapsed

If you have the LWC test pane expanded as well, it only exacerbates the virtual space issue

**Comment by peternhale** [comment]
> @peternhale I've updated that. I will probably try to later contribute a piece of configuration to allow consumers to set the default state for tests, though, as expanded by default makes for a lot of scrolling in a project with more than a few tests, and particularly when there are more than a few tests in the first test class shown.
> 
> Here are a few examples using 2560x1440 resolution:
> 
> * [Apex Rollup](https://github.com/jamessimone/apex-rollup): shows 2.5 test classes with all tests expanded, shows all 30 test classes collapsed
> * [Nebula Logger](https://nebulalogger.com/): shows nearly 7 test classes expanded, shows 57 out of 65 tests collapsed
> * Internal Salesforce.com repo 1: shows 4 test classes expanded, shows all 24 test classes collapsed
> * Internal Salesforce.com repo 2: shows 2 test classes expanded, shows all 18 test classes collapsed
> 
> If you have the LWC test pane expanded as well, it only exacerbates the virtual space issue

Thanks @jamessimone While I appreciate being at parity with LWC tests (collapsed by default), we will need some time to change our end-to-end tests to account for a new initial state of the apex test tree. Once we have the test in place, then the initial state of collapsed can be enabled.

The other thing to note with parity to LWC tests with this change is the two will not be functionally equal, as the collapse all is not present there. Care to contribute that change as well? If not, we will log a work item to capture that change.

**Comment by jamessimone** [comment]
I can contribute the LWC side of things within this PR, sure.

> we will need some time to change our end-to-end tests to account for a new initial state of the apex test tree

When you're referring to e2e tests, is `packages/salesforcedx-vscode-apex/test/vscode-integration/views/testOutline.test.ts` the test in question? At the moment there aren't any tests there asserting on the initial expanded/collapsed state; I had originally added an assert to `Should create one test and one class`  to validate that `testChildGroup` was now collapsed by default, and I'd be happy to re-add that assert if that's what you're referring to.

**Comment by peternhale** [comment]
@jamessimone Thank you! The E2E tests for vscode extensions are here [salesforcedx-vscode-automation-tests](https://github.com/forcedotcom/salesforcedx-vscode-automation-tests)

**Comment by jamessimone** [comment]
@peternhale I'll circle back around to that repo, then - for now, I've added the LWC functionality as well

**Comment by peternhale** [comment]
@jamessimone thanks for the changes. One thing I noticed is the placement of the collapse icon, in Apex tests it appears last, where in LWC it appears first.

**Comment by jamessimone** [comment]
@peternhale as far as I can tell, VS Code determines the placement by way of alphabetization - is that correct?

 I was following the naming conventions for the pre-existing commands for Apex/LWC:

**LWC**
- SFDX: Lightning Web Component Collapse All Tests
- SFDX: Run All Lightning Web Component Tests
- SFDX: Refresh Lightning Web Component Test Explorer

** Apex **
- Refresh Tests
- Run Tests
- SFDX: Collapse All Apex Tests

My thought was that it would be nice to have the "Collapse" commands prefixed with `SFDX: ` as they're also accessible via the Command Palette, but if I removed the `SFDX:` prefix from both, that would consistently put collapse at the start of each group:
image
image


Thoughts?

**Comment by peternhale** [comment]
> I could prefix the other Apex labels with SFDX as well, which would also serve to standardize the ordering. Let me know!

Thanks for the info @jamessimone, I will bring this up with PM and manager today to see if we can reach consensus.

**Comment by peternhale** [comment]
Hi @jamessimone sorry for the delay. I found a way to enforce order for the menu items.

In package.json for each extension one uses the group property, specifying the order index by adding `@n` to `navigation` value.

In the case LWC it would look like this 
```
      "view/title": [
        {
          "command": "sf.lightning.lwc.test.runAllTests",
          "when": "view == sf.lightning.lwc.test.view",
          "group": "navigation@1"
        },
        {
          "command": "sf.lightning.lwc.test.refreshTestExplorer",
          "when": "view == sf.lightning.lwc.test.view",
          "group": "navigation@2"
        },
        {
          "command": "sf.lightning.lwc.test.view.collapseAll",
          "when": "view == sf.lightning.lwc.test.view",
          "group": "navigation@3"
        }
      ],
```

**Comment by jamessimone** [comment]
@peternhale nice find, thank you for tracking that down! I can make that change

**Review by peternhale — APPROVED** [review]

**Comment by peternhale** [comment]
@jamessimone would you mind adding ` @W-16273375@` under section `What issues does this PR fix or reference?` of the commit message?

**Comment by jamessimone** [comment]
I added the WI number, thanks @peternhale !

**Pull request merged**